poc/icart_hacspec.py

```diff
@@ -0,0 +1,57 @@
+from hacspec.speclib import *
+
+prime = 2**384 - 2**128 - 2**96 + 2**32 - 1
+
+felem_t = refine(nat, lambda x: x < prime)
+affine_t = tuple2(felem_t, felem_t)
+
+@typechecked
+def to_felem(x: nat_t) -> felem_t:
+    return felem_t(nat(x % prime))
+
+
+@typechecked
+def fadd(x: felem_t, y: felem_t) -> felem_t:
+    return to_felem(x + y)
+
+
+@typechecked
+def fsub(x: felem_t, y: felem_t) -> felem_t:
+    return to_felem(x - y)
+
+
+@typechecked
+def fmul(x: felem_t, y: felem_t) -> felem_t:
+    return to_felem(x * y)
+
+
+@typechecked
+def fsqr(x: felem_t) -> felem_t:
+    return to_felem(x * x)
+
+
+@typechecked
+def fexp(x: felem_t, n: nat_t) -> felem_t:
+    return to_felem(pow(x, n, prime))
+
+
+@typechecked
+def finv(x: felem_t) -> felem_t:
+    return to_felem(pow(x, prime-2, prime))
+
+a384 = to_felem(prime - 3)
+b384 = to_felem(27580193559959705877849011840389048093056905856361568521428707301988689241309860865136260764883745107765439761230575)
+
+def map2p384(u:felem_t) -> affine_t:
```

Comment:
All functions must be annotated `@typechecked` to pass the hacspec checker.

Reply:
Aha! Thanks. :-)